### ISSUE TITLE
Unexpected behavior of fitness weights, values and weighted values

### ISSUE THREAD

**Issue by PaulinaDabrowska** [DEAP/deap#390]
tagging @morales-gregorio 

We started to learn to use DEAP some time ago and we found several surprising and/or incongruent behaviors of base.Fitness objects. 

First of all, it is possible to assign different number of fitness values and weights, which results in a number of weighted fitnesses to be the same as the shorter of them. Since only weighted values (wvalues) are considered during comparisons of individuals, this may lead to logical errors difficult to track. We would suggest at least a warning, if not an assertion that all values have their corresponding weights as soon as both are defined. 

Related to this, it is possible to set weight values anew at any time. It seems unproper for two reasons: 
1) initial weights are typically defined at the level of class generation with deap.creator, and they are the only feature different between multiple such classes, so modifying them later at the level of particular individuals effectively changes the fitness class of this individual, and consequently the individual itself (from this point on we cannot assume, that the individual shares all features characteristic for its class with other individuals of the same class)
2) depending on whether values were already defined, reassignment of weights has different effect: if no fitness values are defined, modification of weights will only modify weights; if both weights and values (and therefore wvalues) are already defined, modification of weights will lead to modification of values so as to preserve wvalues unmodified. I find this very counterintuitive. Here is the code showing this behavior:
```
import random
from deap import base, creator, tools

creator.create('Fit', base.Fitness, weights=(1.0, 1.0, 1.0))
creator.create('Individual', list, fitness=creator.Fit)

toolbox = base.Toolbox()
toolbox.register("attr", random.randint, 0, 1)
toolbox.register("individual", tools.initRepeat, creator.Individual, toolbox.attr, 10)

i1 = toolbox.individual()
i1.fitness.values = (1.0, 1.0, 1.0)
print(i1.fitness.weights, i1.fitness.values, i1.fitness.wvalues, i1.fitness.valid)
i1.fitness.weights = (1.0, 0.5, 1.0)
print(i1.fitness.weights, i1.fitness.values, i1.fitness.wvalues, i1.fitness.valid)
```
results in:
```
>>(1.0, 1.0, 1.0) (1.0, 1.0, 1.0) (1.0, 1.0, 1.0) True
>>(1.0, 0.5, 1.0) (1.0, 2.0, 1.0) (1.0, 1.0, 1.0) True
```
Our point is that making such changes possible causes not user-robust code and less experienced users may easily get into trouble. 

The last unexpected behavior we observed is related to difference between integers and floats. If both weight and corresponding value are given as integers, the resulting wvalue will be an integer, even if all other weights, values and wvalues are floats. Surprisingly though, if one checks the type of these numbers after assignment, value is a float (weight and wvalue are ints). This would not really matter, if the comparison of individuals gave correct results:
```
# continuing previous code snippet
i4 = toolbox.individual()
i4.fitness.values = (1.0, 1.0, 1.0)
i5 = toolbox.individual()
i5.fitness.weights = (1, 1.0, 1.0)
i5.fitness.values = (1, 1.0, 1.0)

print('compare float and integer fitness values')
for i in [i4, i5]:
    print(i.fitness.weights, i.fitness.values, i.fitness.wvalues, i.fitness.valid)
print('i4 == i5', i4 == i5)
print('i4 > i5', i4 > i5)
```
```
>>compare float and integer fitness values
>>(1.0, 1.0, 1.0) (1.0, 1.0, 1.0) (1.0, 1.0, 1.0) True
>>(1, 1.0, 1.0) (1.0, 1.0, 1.0) (1, 1.0, 1.0) True
>>i4 == i5 False
>>i4 > i5 True
```
We suppose it has to do with floating point precision and/or lexicographical comparisons of tuples, but we did not manage to reproduce this behavior with raw integers, floats and tuples of them. 

What we discussed here is partially related to issue #377.

**Comment by fmder** [comment]
It is probably a typo but I just want to make sure, when you compare individuals, you should compare their fitness with ``i1.fitness == i2.fitness``. Is it the case?

**Comment by fmder** [comment]
I agree the behaviour of weights is too complex for the benefits. However, in Python it is hard to limit users in what they can do. It is assumed that developers know what they are doing when they access a variables. I'm not sure I would prevent anyone changing the weights, but it is a good idea to test if the values are smaller than the weights.

What behaviour would you expect?

**Comment by morales-gregorio** [comment]
tagging @PaulinaDabrowska 

To first address the comparison using `ind.fitness` we have rerun the snippet:
```
# continuing previous code snippet
i4 = toolbox.individual()
i4.fitness.values = (1.0, 1.0, 1.0)
i5 = toolbox.individual()
i5.fitness.weights = (1, 1.0, 1.0)
i5.fitness.values = (1, 1.0, 1.0)

print('compare float and integer fitness values')
for i in [i4, i5]:
    print(i.fitness.weights, i.fitness.values,
          i.fitness.wvalues, i.fitness.valid)
print('i4 == i5', i4.fitness == i5.fitness)
print('i4 > i5', i4.fitness > i5.fitness)
```

Which has fixed the strange behavior of the comparisons:
```
>>compare float and integer fitness values
>>(1.0, 1.0, 1.0) (1.0, 1.0, 1.0) (1.0, 1.0, 1.0) True
>>(1, 1.0, 1.0) (1.0, 1.0, 1.0) (1, 1.0, 1.0) True
>>i4 == i5 True
>>i4 > i5 False
```

But this raises the question, what is the comparison `i1 == i2` used for? Since if the two individuals with same parameters are considered unequal.

**Comment by morales-gregorio** [comment]
As for the weight reassignment. 
We would expect that the weights are modified and new fitness.wvalues are recalculated, while leaving the fitness.values unchanged.

**Comment by PaulinaDabrowska** [comment]
I am sorry, it was my mistake to compare individuals instead of individuals.fitness. Nevertheless we have just examined the behavior of corrected code and found out, that `i1.fitness == i2.fitness` compares fitness.values, not fitness.wvalues, so that weights are not used. In case compared individuals have different weights, this comparison has no sense. 

Regarding the reassignments of weights, as @morales-gregorio wrote, we would expect that wvalues are then recalculated, leaving values unchanged.

**Comment by PaulinaDabrowska** [comment]
And regarding the size of values and weights, I would add sth like `assert(len(weights) == len(values)` wherever one or the other are assigned, with a meaningful assertion error message stating, that they should be of equal length.

**Comment by fmder** [comment]
Values is a property of fitness. If you look at [base.py](https://github.com/DEAP/deap/blob/a5ec46e48dd384f99586387f754c8a43279b3d5c/deap/base.py#L235) the wvalues are compared.

I agree with asserting.

**Comment by cyasing** [comment]
Can anyone explain to me how to resolve this error I am getting for FitnessMulti type with weights=(-1.0, 1.0) and 5 variables to be optimized? Error: `"Assigned values have not the same length than fitness weights"`
